packages/@atjson/editor/src/text-input.ts

```diff
@@ -0,0 +1,277 @@
+
+  beforeinput(evt: InputEvent) {
+
+    if (evt.isComposing) return;
+
+    let ranges = evt.getTargetRanges();
+    let { start, end } = this.selection;
+
+    console.log('here is a thing', evt);
+
+    switch (evt.inputType) {
+    case 'insertText':
+      if (start === end) {
+        this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: evt.data } }));
+      } else {
+        this.dispatchEvent(new CustomEvent('replaceText', { detail: { start, end, text: evt.data } }));
+      }
+      break;
+
+    case 'insertLineBreak':
+      this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: '\u2028' } }));
+      this.dispatchEvent(new CustomEvent('addAnnotation', {
+        detail: { type: 'line-break', start, end: end + 1 }
+      }));
+      break;
+
+    case 'insertFromPaste':
+      var text = evt.dataTransfer.getData('text/plain');
+
+      if (start === end) {
+        this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: text } }));
+      } else {
+        this.dispatchEvent(new CustomEvent('replaceText', { detail: { start: start, end: end, text: text } }));
+      }
+      break;
+
+    case 'insertFromDrop':
+      var text = evt.dataTransfer.getData('text/plain');
+
+      var target = evt.getTargetRanges()[0];
+      start = this.nodeAndOffsetToDocumentOffset(target.startContainer, target.startOffset);
+
+      this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: text } }));
+
+      break;
+
+
+    case 'deleteContentBackward':
+      if (this.selection.collapsed) {
+        start--;
+      }
+      this.dispatchEvent(new CustomEvent('deleteText', {
+        detail: { start, end }
+      }));
+      break;
+
+    case 'deleteWordBackward':
+    case 'deleteWordForward':
+      if (evt.inputType === 'deleteWordBackward' && this.selection.collapsed) {
+        end++;
+      }
+
+      var target = evt.getTargetRanges()[0];
+      let deletionStart = this.nodeAndOffsetToDocumentOffset(target.startContainer, target.startOffset);
+      let deletionEnd = this.nodeAndOffsetToDocumentOffset(target.endContainer, target.endOffset);
+
+      this.dispatchEvent(new CustomEvent('deleteText', {
+        detail: { start: deletionStart, end: deletionEnd }
+      }));
+      break;
+
+    case 'deleteContentForward':
+      if (this.selection.collapsed) {
+        end++;
+      }
+      this.dispatchEvent(new CustomEvent('deleteText', {
+        detail: { start, end }
+      }));
+      break;
+
+    case 'deleteByCut':
+    case 'deleteContent':
+    case 'deleteByDrag':
+      this.dispatchEvent(new CustomEvent('deleteText', {
+        detail: { start, end }
+      }));
+      break;
+
+    case 'insertReplacementText':
+      console.log('in here with evt', evt);
+      // n.b. this assumes that there is only one target range.
+      if (evt.getTargetRanges().length !== 1) {
```

Comment:
Urgh, could you leave a comment and a link to docs that shows how we can get multiple target ranges?

Also, a use-case here would be ✨ lovely ✨  (I'm assuming the primary use case is find & replace)